### PR TITLE
plugins: implement plugin zip loading

### PR DESCRIPTION
**Pull request by nikosszzz**
* Add fallback screen for the plugin page when no plugins are installed
* Add plugin manifest support
* Add /plugins command
* Add a better coming soon page to Updater and Themer
* Add Toast alert when enabling/disabling plugins

Signed-off-by: nikosszzz <19nikospap@gmail.com>

### CODE REVIEW

File: src/api/PluginManager.ts [hunk 1]
```diff
@@ -54,10 +73,10 @@ export function startCorePlugins() {
     for (const pluginClass of [Badges, CommandHandler, CoreCommands, NoTrack]) {
         const { name } = pluginClass;
         try {
-            logger.info("Loading CorePlugin " + name);
+            logger.info("Loading CorePlugin: " + name);
```
Comment (Vendicated) [inline src/api/PluginManager.ts:76]: but why

File: src/api/Settings.ts [hunk 2]
```diff
@@ -18,14 +18,12 @@ type CastDown<T> =
  * save to settings on set()
  */
 export function useSettings<T>(settings: Settings<T>) {
-    const [, update] = React.useState(0);
```
Comment (Vendicated) [inline src/api/Settings.ts:21]: add this back

File: src/api/PluginManager.ts (hunk 1 above)
Comment (C10udburst) [reply src/api/PluginManager.ts:76]: why not

File: src/entities/types.ts [hunk 3]
```diff
@@ -2,10 +2,12 @@ export interface PluginManifest {
     name: string;
     description: string;
     version: string;
-    authors: PluginAuthor[];
-}
+    authors: PluginAuthor;
+}[];
```
Comment (Vendicated) [inline src/entities/types.ts:6]: why did you make manifest an array?

File: src/api/PluginManager.ts [hunk 4]
```diff
@@ -40,24 +39,25 @@ export async function startPlugins() {
             const zip = new ZipFile(PLUGINS_DIRECTORY + file.name, 0, "r");
             try {
                 zip.openEntry("manifest.json");
-                const manifestBuffer = JSON.parse(zip.readEntry("text")) as any;
+                const manifest = JSON.parse(zip.readEntry("text"));
                 zip.closeEntry();
 
                 zip.openEntry("index.js.bundle");
                 const pluginBuffer = zip.readEntry("binary");
                 zip.closeEntry();
```
Comment (Vendicated) [inline src/api/PluginManager.ts:47]: move this below the checks to not needlessly load it

File: src/api/PluginManager.ts [hunk 5]
```diff
@@ -40,24 +39,25 @@ export async function startPlugins() {
...
                 try {
-                    logger.info(`Loading Plugin ${name}...`);
-                    const plugin = plugins[name] = new pluginClass(new Settings(name));
-                    plugin.manifest = manifestBuffer;
-                    if (isPluginEnabled(plugin.name)) plugin.start();
+                    if (pluginClass.prototype instanceof Plugin) {
+                        if (manifest.name !== pluginClass.name) throw new Error(`Plugin ${manifest.name} must export a class named ${manifest.name}`);
+                        logger.info(`Loading Plugin ${manifest.name}...`);
+                        const plugin = plugins[manifest] = new pluginClass(manifest);
+                        if (isPluginEnabled(plugin.name)) plugin.start();
```
Comment (Vendicated) [inline src/api/PluginManager.ts:57]: remove this line

Comment (Vendicated) [reply src/api/PluginManager.ts:57]:
```suggestion
                        plugin.start();
```

**Review by Vendicated — APPROVED** [review]

**Pull request merged**